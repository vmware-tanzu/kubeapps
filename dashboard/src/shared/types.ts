// Copyright 2018-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { JSONSchemaType } from "ajv";
import { RouterState } from "connected-react-router";
import {
  AvailablePackageDetail,
  AvailablePackageSummary,
  GetAvailablePackageSummariesResponse,
  InstalledPackageDetail,
  InstalledPackageSummary,
  PackageAppVersion,
  ResourceRef,
} from "gen/kubeappsapis/core/packages/v1alpha1/packages";
import {
  DockerCredentials,
  OpaqueCredentials,
  PackageRepositoryAuth_PackageRepositoryAuthType,
  SshCredentials,
  TlsCertKey,
  UsernamePassword,
} from "gen/kubeappsapis/core/packages/v1alpha1/repositories";
import { Plugin } from "gen/kubeappsapis/core/plugins/v1alpha1/plugins";
import { HelmPackageRepositoryCustomDetail } from "gen/kubeappsapis/plugins/helm/packages/v1alpha1/helm";
import { KappControllerPackageRepositoryCustomDetail } from "gen/kubeappsapis/plugins/kapp_controller/packages/v1alpha1/kapp_controller";
import { IOperatorsState } from "reducers/operators";
import { Subscription } from "rxjs";
import { IAuthState } from "../reducers/auth";
import { IClustersState } from "../reducers/cluster";
import { IConfigState } from "../reducers/config";
import { IPackageRepositoryState } from "../reducers/repos";
import { RpcError } from "./RpcError";

export class CustomError extends Error {
  public causes: Error[] | undefined;
  // The constructor is defined so we can later on compare the returned object
  // via err.constructor  == FOO
  constructor(message?: string, causes?: Error[]) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
    this.causes = causes;
    this.checkCauses();
  }
  // Workaround used until RPC code (unary) throws a custom rpc error
  // Check if any RPC error is among the causes
  private checkCauses() {
    if (!this.causes) return;
    for (let i = 0; i < this.causes.length; i++) {
      const cause = this.causes[i];
      if (RpcError.isRpcError(cause)) {
        this.causes[i] = new RpcError(cause);
      }
    }
  }
}

// For 4XX HTTP-alike errors
export class ClientNetworkError extends CustomError {}
// 400
export class BadRequestNetworkError extends ClientNetworkError {}
// 401
export class UnauthorizedNetworkError extends ClientNetworkError {}
// 403
export class ForbiddenNetworkError extends ClientNetworkError {}
// 404
export class NotFoundNetworkError extends ClientNetworkError {}
// 408
export class RequestTimeoutNetworkError extends ClientNetworkError {}
// 409
export class ConflictNetworkError extends ClientNetworkError {}
// 422
export class UnprocessableEntityError extends ClientNetworkError {}
// 429
export class TooManyRequestsNetworkError extends ClientNetworkError {}

// For 5XX HTTP-alike errors
export class ServerNetworkError extends CustomError {}
// 500
export class InternalServerNetworkError extends ServerNetworkError {}
// 501
export class NotImplementedNetworkError extends ServerNetworkError {}
// 503
export class ServerUnavailableNetworkError extends ServerNetworkError {}
// 504
export class GatewayTimeoutNetworkError extends ServerNetworkError {}

// Application-level errors
export class FetchError extends CustomError {}

export class FetchWarning extends CustomError {}

export class CreateError extends CustomError {}

export class UpgradeError extends CustomError {}

export class RollbackError extends CustomError {}

export class DeleteError extends CustomError {}

export type DeploymentEvent = "install" | "upgrade";

export interface IReceivePackagesActionPayload {
  response: GetAvailablePackageSummariesResponse;
  paginationToken: string;
}

export interface IPackageState {
  isFetching: boolean;
  hasFinishedFetching: boolean;
  nextPageToken: string;
  selected: {
    error?: FetchError | Error;
    availablePackageDetail?: AvailablePackageDetail;
    pkgVersion?: string;
    appVersion?: string;
    versions: PackageAppVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
    schema?: JSONSchemaType<any>;
  };
  items: AvailablePackageSummary[];
  categories: string[];
  size: number;
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip?: string; hostname?: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IHTTPIngressPath {
  path: string;
}

export interface IIngressHTTP {
  paths: IHTTPIngressPath[];
}

export interface IIngressRule {
  host: string;
  http: IIngressHTTP;
}

export interface IIngressTLS {
  hosts: string[];
}

export interface IIngressSpec {
  rules: IIngressRule[];
  tls?: IIngressTLS[];
  backend?: {
    serviceName: string;
    servicePort: number;
  };
}

export interface IResourceMetadata {
  name: string;
  namespace: string;
  annotations: { [key: string]: string };
  ownerReferences?: Array<{
    apiVersion: string;
    blockOwnerDeletion: string;
    kind: string;
    name: string;
    uid: string;
  }>;
  creationTimestamp: string;
  selfLink: string;
  resourceVersion: string;
  deletionTimestamp?: string;
  uid: string;
}

export interface IResource {
  apiVersion: string;
  kind: string;
  type: string;
  spec: any;
  status: any;
  metadata: IResourceMetadata;
}

export interface ISecret {
  apiVersion: string;
  kind: string;
  type: string;
  data: { [s: string]: string };
  metadata: IResourceMetadata;
}

export interface IPackageManifestChannel {
  name: string;
  currentCSV: string;
  currentCSVDesc: {
    annotations: {
      "alm-examples": string;
      capabilities: string;
      categories: string;
      certified: string;
      containerImage: string;
      createdAt: string;
      description: string;
      repository: string;
      support: string;
    };
    description: string;
    displayName: string;
    provider: {
      name: string;
    };
    version: string;
    installModes: [
      { supported: boolean; type: "OwnNamespace" },
      { supported: boolean; type: "SingleNamespace" },
      { supported: boolean; type: "MultiNamespace" },
      { supported: boolean; type: "AllNamespaces" },
    ];
    customresourcedefinitions: {
      owned: Array<{
        description: string;
        displayName: string;
        kind: string;
        name: string;
        version: string;
      }>;
    };
  };
}

export interface IPackageManifestStatus {
  catalogSource: string;
  catalogSourceDisplayName: string;
  catalogSourceNamespace: string;
  catalogSourcePublisher: string;
  provider: {
    name: string;
  };
  defaultChannel: string;
  channels: IPackageManifestChannel[];
}

export interface IPackageManifest extends IResource {
  status: IPackageManifestStatus;
}

export interface IClusterServiceVersionCRDResource {
  kind: string;
  name: string;
  version: string;
}

export interface IClusterServiceVersionCRD {
  description: string;
  displayName: string;
  kind: string;
  name: string;
  version: string;
  resources: IClusterServiceVersionCRDResource[];
  specDescriptors: Array<{
    description: string;
    displayName: string;
    path: string;
    "x-descriptors": string[];
  }>;
  statusDescriptors: Array<{
    description: string;
    displayName: string;
    path: string;
    "x-descriptors": string[];
  }>;
}

export interface IClusterServiceVersionSpec {
  apiservicedefinitions: any;
  customresourcedefinitions: {
    owned?: IClusterServiceVersionCRD[];
  };
  description: string;
  displayName: string;
  icon: Array<{
    base64data: string;
    mediatype: string;
  }>;
  install: any;
  installModes: [
    { supported: boolean; type: "OwnNamespace" },
    { supported: boolean; type: "SingleNamespace" },
    { supported: boolean; type: "MultiNamespace" },
    { supported: boolean; type: "AllNamespaces" },
  ];
  keywords: string[];
  labels: any;
  links: Array<{
    name: string;
    url: string;
  }>;
  maintainers: Array<{
    email: string;
    name: string;
  }>;
  maturity: string;
  provider: {
    name: string;
  };
  selector: any;
  version: string;
}

export interface IClusterServiceVersion extends IResource {
  spec: IClusterServiceVersionSpec;
}

export interface IPkgRepositoryFilter {
  jq: string;
  variables?: { [key: string]: string };
}

export interface IInstalledPackageState {
  isFetching: boolean;
  error?: FetchError | CreateError | UpgradeError | RollbackError | DeleteError;
  items: InstalledPackageDetail[];
  listOverview?: InstalledPackageSummary[];
  selected?: CustomInstalledPackageDetail;
  // TODO(agamez): add tests for this new state field
  selectedDetails?: AvailablePackageDetail;
  resourceRefs?: ResourceRef[];
}

export interface IStoreState {
  router: RouterState;
  apps: IInstalledPackageState;
  auth: IAuthState;
  packages: IPackageState;
  config: IConfigState;
  kube: IKubeState;
  repos: IPackageRepositoryState;
  clusters: IClustersState;
  operators: IOperatorsState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IKubeItem<T> {
  isFetching: boolean;
  item?: T;
  error?: Error;
}

export interface IKind {
  apiVersion: string;
  plural: string;
  namespaced: boolean;
}

export interface IKubeState {
  items: { [s: string]: IKubeItem<IResource | IK8sList<IResource, {}>> };
  subscriptions: { [s: string]: Subscription };
  // TODO(minelson): Remove kinds and kindsError once the operator support is
  // removed from the dashboard or replaced with a plugin.
  kinds: { [kind: string]: IKind };
  kindsError?: Error;
}

export interface IBasicFormParam {
  path: string;
  type?: "string" | "number" | "integer" | "boolean" | "object" | "array" | "null" | "any";
  value?: any;
  title?: string;
  minimum?: number;
  maximum?: number;
  render?: string;
  description?: string;
  customComponent?: object;
  enum?: string[];
  hidden?:
    | {
        event: DeploymentEvent;
        path: string;
        value: string;
        conditions: Array<{
          event: DeploymentEvent;
          path: string;
          value: string;
        }>;
        operator: string;
      }
    | string;
  children?: IBasicFormParam[];
}
export interface IBasicFormSliderParam extends IBasicFormParam {
  sliderMin?: number;
  sliderMax?: number;
  sliderStep?: number;
  sliderUnit?: string;
}

export interface CustomInstalledPackageDetail extends InstalledPackageDetail {
  revision: number;
}

//  enum for the type of package repository storage
export enum RepositoryStorageTypes {
  PACKAGE_REPOSITORY_STORAGE_HELM = "helm",
  PACKAGE_REPOSITORY_STORAGE_OCI = "oci",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_INLINE = "inline",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_IMAGE = "image",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_IMGPKGBUNDLE = "imgpkgBundle",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_HTTP = "http",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_GIT = "git",
}

export enum PluginNames {
  PACKAGES_HELM = "helm.packages",
  PACKAGES_FLUX = "fluxv2.packages",
  PACKAGES_KAPP = "kapp_controller.packages",
}

export interface IPkgRepoFormData {
  authMethod: PackageRepositoryAuth_PackageRepositoryAuthType;
  // kubeapps-managed secrets
  authHeader: string; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_BEARER || type == PACKAGE_REPOSITORY_AUTH_TYPE_AUTHORIZATION_HEADER
  basicAuth: UsernamePassword; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_BASIC_AUTH
  dockerRegCreds: DockerCredentials; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_DOCKER_CONFIG_JSON
  sshCreds: SshCredentials; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_SSH
  opaqueCreds: OpaqueCredentials; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_OPAQUE
  tlsCertKey: TlsCertKey; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_TLS
  // user-managed secrets
  secretAuthName: string;
  secretTLSName: string;
  // rest of the parameters
  customCA: string;
  description: string;
  interval: string;
  name: string;
  passCredentials: boolean;
  plugin: Plugin;
  skipTLS: boolean;
  type: string;
  url: string;
  // add more types if necessary
  customDetail?: Partial<
    HelmPackageRepositoryCustomDetail | KappControllerPackageRepositoryCustomDetail
  >;
  namespace: string;
  isNamespaceScoped: boolean;
}
