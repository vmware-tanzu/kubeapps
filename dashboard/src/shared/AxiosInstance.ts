// Copyright 2018-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import Axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from "axios";
import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictNetworkError,
  ForbiddenNetworkError,
  InternalServerNetworkError,
  IRBACRole,
  IStoreState,
  NotFoundNetworkError,
  UnauthorizedNetworkError,
  UnprocessableEntityError,
} from "./types";

export function addAuthHeaders(axiosInstance: AxiosInstance) {
  axiosInstance.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken && config?.headers) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
}

export function addErrorHandling(axiosInstance: AxiosInstance, store: Store<IStoreState>) {
  axiosInstance.interceptors.response.use(
    response => response,
    e => {
      const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
      const err: AxiosError<any, any> = e;
      if (
        err.code === undefined &&
        err.message === "Network Error" &&
        !err.response &&
        Auth.usingOIDCToken()
      ) {
        // The OIDC token is no longer valid, logout
        dispatch(actions.auth.expireSession());
      }
      let message = err.message;
      if (err.response) {
        if (err.response.data.message) {
          message = err.response.data.message;
        }
        if (typeof err.response.data === "string") {
          message = err.response.data;
        }
      }

      const dispatchErrorAndLogout = (m: string) => {
        // Global action dispatch to log the user out
        dispatch(actions.auth.authenticationError(m));
        // Expire the session if we are using OIDC tokens and
        // logout either way.
        dispatch(actions.auth.expireSession());
      };
      const response = err.response as AxiosResponse<any>;
      switch (response && response.status) {
        case 401:
          dispatchErrorAndLogout(message);
          return Promise.reject(new UnauthorizedNetworkError(message));
        case 403:
          // Subcase 1:
          //   if usingOIDCToken: a 403 directly from the auth proxy
          //   always requires reauthentication.
          // Subcase 2:
          //   if !usingOIDCToken, an anonymous response is sent when
          //   a serviceaccount token expires (eg., delete secret xxx)
          //   or the token is managed externally (eg., vsphere kubectl login)
          //   In this case, force reauthentication
          if (Auth.is403FromAuthProxy(response) || Auth.isAnonymous(response)) {
            dispatchErrorAndLogout(message);
          }
          // Subcase 3:
          //   The most likely case is just a 403 due to a lack of
          //   permissions in a certain namespace.
          //   In this case, just return the error message back to the user
          try {
            const jsonMessage = JSON.parse(message) as IRBACRole[];
            return Promise.reject(
              new ForbiddenNetworkError(
                `Forbidden error, missing permissions: ${jsonMessage
                  .map(forbiddenAction => {
                    const { apiGroup, resource, namespace, clusterWide, verbs } = forbiddenAction;
                    return `apiGroup: "${apiGroup}", resource: "${resource}", action: "${verbs.join(
                      ", ",
                    )}", ${clusterWide ? "in all namespaces" : `namespace: ${namespace}`}`;
                  })
                  .join("; ")}`,
              ),
            );
          } catch (e: any) {
            // Subcase 4:
            //   A non-parseable 403 error.
            //   Do not require reauthentication and display error (ie. edge cases of proxy auth)
          }
          return Promise.reject(new ForbiddenNetworkError(message));
        case 404:
          return Promise.reject(new NotFoundNetworkError(message));
        case 409:
          return Promise.reject(new ConflictNetworkError(message));
        case 422:
          return Promise.reject(new UnprocessableEntityError(message));
        case 500:
          return Promise.reject(new InternalServerNetworkError(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}

// Error handling is added with an interceptor in index.tsx
export const axios = Axios.create();
// Authorization headers and error handling are added with an interceptor in index.tsx
export const axiosWithAuth = Axios.create();
