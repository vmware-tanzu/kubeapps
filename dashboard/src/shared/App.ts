import {
  AvailablePackageDetail,
  AvailablePackageReference,
  Context,
  CreateInstalledPackageRequest,
  InstalledPackageReference,
  ReconciliationOptions,
  VersionReference,
} from "gen/kubeappsapis/core/packages/v1alpha1/packages";
import * as url from "shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { KubeappsGrpcClient } from "./KubeappsGrpcClient";

export const KUBEOPS_ROOT_URL = "api/kubeops/v1";
export class App {
  private static client = () => new KubeappsGrpcClient().getPackagesServiceClientImpl();

  public static async GetInstalledPackageSummaries(
    cluster: string,
    namespace?: string,
    page?: number,
    size?: number,
  ) {
    return await this.client().GetInstalledPackageSummaries({
      context: { cluster: cluster, namespace: namespace },
      paginationOptions: { pageSize: size || 0, pageToken: page?.toString() || "0" },
    });
  }

  public static async GetInstalledPackageDetail(installedPackageRef?: InstalledPackageReference) {
    return await this.client().GetInstalledPackageDetail({
      installedPackageRef: installedPackageRef,
    });
  }

  public static async createInstalledPackage(
    targetContext: Context,
    name: string,
    availablePackageRef: AvailablePackageReference,
    pkgVersionReference: VersionReference,
    values?: string,
    reconciliationOptions?: ReconciliationOptions,
  ) {
    return await this.client().CreateInstalledPackage({
      name,
      values,
      targetContext,
      availablePackageRef,
      pkgVersionReference,
      reconciliationOptions,
    } as CreateInstalledPackageRequest);
  }

  public static async upgrade(
    installedPackageRef: InstalledPackageReference,
    chartNamespace: string,
    availablePackageDetail: AvailablePackageDetail,
    values?: string,
  ) {
    const endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName:
        availablePackageDetail.availablePackageRef?.identifier.split("/")[0],
      appRepositoryResourceNamespace: chartNamespace,
      chartName: decodeURIComponent(availablePackageDetail.name),
      releaseName: installedPackageRef.identifier,
      values,
      version: availablePackageDetail.version?.pkgVersion,
    });
    return data;
  }

  public static async rollback(installedPackageRef: InstalledPackageReference, revision: number) {
    const endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(installedPackageRef: InstalledPackageReference, purge: boolean) {
    let endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    if (purge) {
      endpoint += "?purge=true";
    }
    const { data } = await axiosWithAuth.delete(endpoint);
    return data;
  }

  // TODO(agamez): remove it once we return the generated resources as part of the InstalledPackageDetail.
  public static async getRelease(installedPackageRef?: InstalledPackageReference) {
    const { data } = await axiosWithAuth.get<{ data: { manifest: any } }>(
      url.kubeops.releases.get(
        installedPackageRef?.context?.cluster ?? "",
        installedPackageRef?.context?.namespace ?? "",
        installedPackageRef?.identifier ?? "",
      ),
    );
    return data.data;
  }
}
