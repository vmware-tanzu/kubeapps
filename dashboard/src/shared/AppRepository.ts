import { axiosWithAuth } from "./AxiosInstance";
import { IAppRepositoryFilter, ICreateAppRepositoryResponse } from "./types";
import * as url from "./url";

export class AppRepository {
  public static async list(cluster: string, namespace: string) {
    const {
      data: { appRepository },
    } = await axiosWithAuth.get<any>(url.backend.apprepositories.list(cluster, namespace));
    return appRepository;
  }

  public static async get(cluster: string, namespace: string, name: string) {
    const {
      data: { appRepository },
    } = await axiosWithAuth.get<any>(url.backend.apprepositories.get(cluster, namespace, name));
    return appRepository;
  }

  public static async getSecretForRepo(cluster: string, namespace: string, name: string) {
    const {
      data: { secret },
    } = await axiosWithAuth.get<any>(url.backend.apprepositories.get(cluster, namespace, name));
    return secret;
  }

  public static async resync(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.post(
      url.backend.apprepositories.refresh(cluster, namespace, name),
      null,
    );
    return data;
  }

  public static async update(
    cluster: string,
    name: string,
    namespace: string,
    repoURL: string,
    type: string,
    description: string,
    authHeader: string,
    authRegCreds: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
    ociRepositories: string[],
    skipTLS: boolean,
    passCredentials: boolean,
    filter?: IAppRepositoryFilter,
  ) {
    const { data } = await axiosWithAuth.put<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.update(cluster, namespace, name),
      {
        appRepository: {
          name,
          repoURL,
          type,
          description,
          authHeader,
          authRegCreds,
          customCA,
          syncJobPodTemplate,
          registrySecrets,
          ociRepositories,
          tlsInsecureSkipVerify: skipTLS,
          passCredentials: passCredentials,
          filterRule: filter,
        },
      },
    );
    return data;
  }

  public static async delete(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.delete(
      url.backend.apprepositories.delete(cluster, namespace, name),
    );
    return data;
  }

  // create uses the kubeapps backend API
  // TODO(mnelson) Update other endpoints to similarly use the backend API, removing the need
  // for direct k8s api access (for this resource, at least).
  public static async create(
    cluster: string,
    name: string,
    namespace: string,
    repoURL: string,
    type: string,
    description: string,
    authHeader: string,
    authRegCreds: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
    ociRepositories: string[],
    skipTLS: boolean,
    passCredentials: boolean,
    filter?: IAppRepositoryFilter,
  ) {
    const { data } = await axiosWithAuth.post<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.create(cluster, namespace),
      {
        appRepository: {
          name,
          repoURL,
          authHeader,
          authRegCreds,
          type,
          description,
          customCA,
          syncJobPodTemplate,
          registrySecrets,
          ociRepositories,
          tlsInsecureSkipVerify: skipTLS,
          passCredentials: passCredentials,
          filterRule: filter,
        },
      },
    );
    return data;
  }

  public static async validate(
    cluster: string,
    namespace: string,
    repoURL: string,
    type: string,
    authHeader: string,
    authRegCreds: string,
    customCA: string,
    ociRepositories: string[],
    skipTLS: boolean,
    passCredentials: boolean,
  ) {
    const { data } = await axiosWithAuth.post<any>(
      url.backend.apprepositories.validate(cluster, namespace),
      {
        appRepository: {
          repoURL,
          type,
          authHeader,
          authRegCreds,
          customCA,
          ociRepositories,
          tlsInsecureSkipVerify: skipTLS,
          passCredentials: passCredentials,
        },
      },
    );
    return data;
  }
}
