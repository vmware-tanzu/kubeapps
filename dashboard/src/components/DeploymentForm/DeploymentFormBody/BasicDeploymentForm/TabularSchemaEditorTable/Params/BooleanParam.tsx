// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { CdsControlMessage } from "@cds/react/forms";
import { CdsToggle, CdsToggleGroup } from "@cds/react/toggle";
import Column from "components/js/Column";
import Row from "components/js/Row";
import { useState } from "react";
import { IBasicFormParam } from "shared/types";
import { getStringValue } from "shared/utils";

export interface IBooleanParamProps {
  id: string;
  label: string;
  param: IBasicFormParam;
  handleBasicFormParamChange: (
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement>) => void;
}
export default function BooleanParam(props: IBooleanParamProps) {
  const { id, label, param, handleBasicFormParamChange } = props;

  const [currentValue, setCurrentValue] = useState(param.currentValue || false);
  const [isValueModified, setIsValueModified] = useState(false);

  const onChange = (e: React.FormEvent<HTMLInputElement>) => {
    // create an event that "getValueFromEvent" can process,
    const event = {
      currentTarget: {
        //convert the boolean "checked" prop to a normal "value" string one
        value: getStringValue(e.currentTarget?.checked),
        type: "checkbox",
      },
    } as React.FormEvent<HTMLInputElement>;
    setCurrentValue(e.currentTarget?.checked);
    setIsValueModified(e.currentTarget?.checked !== param.currentValue);
    handleBasicFormParamChange(param)(event);
  };

  const unsavedMessage = isValueModified ? "Unsaved" : "";

  const isModified =
    isValueModified ||
    (param.currentValue !== param.defaultValue && param.currentValue !== param.deployedValue);

  const input = (
    <CdsToggleGroup id={id + "_group"}>
      <label htmlFor={id + "_group"}>{""}</label>
      <CdsToggle>
        <input
          required={param.isRequired}
          disabled={param.readOnly}
          aria-label={label}
          id={id}
          type="checkbox"
          onChange={onChange}
          checked={currentValue}
        />
        <CdsControlMessage className={isModified ? "italics" : ""}>
          {currentValue ? "true" : "false"}
        </CdsControlMessage>
      </CdsToggle>
      <CdsControlMessage>{unsavedMessage}</CdsControlMessage>
    </CdsToggleGroup>
  );

  return (
    <Row>
      <Column span={10}>{input}</Column>
    </Row>
  );
}
