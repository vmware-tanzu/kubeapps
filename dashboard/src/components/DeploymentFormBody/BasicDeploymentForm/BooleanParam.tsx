// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import Switch from "react-switch";
import { IBasicFormParam } from "shared/types";

export interface IStringParamProps {
  id: string;
  label: string;
  param: IBasicFormParam;
  handleBasicFormParamChange: (
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement>) => void;
}

function BooleanParam({ id, param, label, handleBasicFormParamChange }: IStringParamProps) {
  // handleChange transform the event received by the Switch component to a checkbox event
  const handleChange = (checked: boolean) => {
    const event = {
      currentTarget: { value: String(checked), type: "checkbox", checked },
    } as React.FormEvent<HTMLInputElement>;
    handleBasicFormParamChange(param)(event);
  };

  return (
    <label htmlFor={id}>
      <div>
        <Switch
          height={20}
          width={40}
          id={id}
          onChange={handleChange}
          checked={!!param.value}
          className="react-switch"
          onColor="#5aa220"
          checkedIcon={false}
          uncheckedIcon={false}
        />
        <span className="deployment-form-label">{label}</span>
      </div>
      {param.description && <span className="description">{param.description}</span>}
    </label>
  );
}

export default BooleanParam;
