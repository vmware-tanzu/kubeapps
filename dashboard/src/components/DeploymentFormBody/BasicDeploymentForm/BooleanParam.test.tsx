// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { mount } from "enzyme";
import { IBasicFormParam } from "shared/types";
import BooleanParam from "./BooleanParam";

const param = { path: "enableMetrics", value: true, type: "boolean" } as IBasicFormParam;
const defaultProps = {
  id: "foo",
  label: "Enable Metrics",
  param,
  handleBasicFormParamChange: jest.fn(),
};

it("should render a boolean param with title and description", () => {
  const wrapper = mount(<BooleanParam {...defaultProps} />);
  const s = wrapper.find(".react-switch").first();
  expect(s.prop("checked")).toBe(defaultProps.param.value);
  expect(wrapper).toMatchSnapshot();
});

it("should send a checkbox event to handleBasicFormParamChange", () => {
  const handler = jest.fn();
  const handleBasicFormParamChange = jest.fn().mockReturnValue(handler);
  const wrapper = mount(
    <BooleanParam {...defaultProps} handleBasicFormParamChange={handleBasicFormParamChange} />,
  );
  const s = wrapper.find(".react-switch").first();

  (s.prop("onChange") as any)(false);

  expect(handleBasicFormParamChange.mock.calls[0][0]).toEqual({
    path: "enableMetrics",
    type: "boolean",
    value: true,
  });
  expect(handler.mock.calls[0][0]).toMatchObject({
    currentTarget: { value: "false", type: "checkbox", checked: false },
  });
});
