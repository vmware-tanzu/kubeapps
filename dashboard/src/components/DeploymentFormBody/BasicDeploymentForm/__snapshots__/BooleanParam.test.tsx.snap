// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a boolean param with title and description 1`] = `
<BooleanParam
  handleBasicFormParamChange={[MockFunction]}
  id="foo"
  label="Enable Metrics"
  param={
    Object {
      "path": "enableMetrics",
      "type": "boolean",
      "value": true,
    }
  }
>
  <label
    htmlFor="foo"
  >
    <div>
      <ReactSwitch
        activeBoxShadow="0 0 2px 3px #3bf"
        boxShadow={null}
        checked={true}
        checkedIcon={false}
        className="react-switch"
        disabled={false}
        height={20}
        id="foo"
        offColor="#888"
        offHandleColor="#fff"
        onChange={[Function]}
        onColor="#5aa220"
        onHandleColor="#fff"
        uncheckedIcon={false}
        width={40}
      >
        <div
          className="react-switch"
          style={
            Object {
              "MozTransition": "opacity 0.25s",
              "MozUserSelect": "none",
              "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
              "WebkitTransition": "opacity 0.25s",
              "WebkitUserSelect": "none",
              "borderRadius": 10,
              "direction": "ltr",
              "display": "inline-block",
              "msUserSelect": "none",
              "opacity": 1,
              "position": "relative",
              "textAlign": "left",
              "touchAction": "none",
              "transition": "opacity 0.25s",
              "userSelect": "none",
            }
          }
        >
          <div
            className="react-switch-bg"
            onClick={[Function]}
            onMouseDown={[Function]}
            style={
              Object {
                "MozTransition": "background 0.25s",
                "WebkitTransition": "background 0.25s",
                "background": "#5aa220",
                "borderRadius": 10,
                "cursor": "pointer",
                "height": 20,
                "margin": 0,
                "position": "relative",
                "transition": "background 0.25s",
                "width": 40,
              }
            }
          />
          <div
            className="react-switch-handle"
            onClick={[Function]}
            onMouseDown={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "MozTransition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
                "WebkitTransition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
                "background": "#ffffff",
                "border": 0,
                "borderRadius": "50%",
                "boxShadow": null,
                "cursor": "pointer",
                "display": "inline-block",
                "height": 18,
                "outline": 0,
                "position": "absolute",
                "top": 1,
                "transform": "translateX(21px)",
                "transition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
                "width": 18,
              }
            }
          />
          <input
            aria-checked={true}
            checked={true}
            disabled={false}
            id="foo"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            role="switch"
            style={
              Object {
                "border": 0,
                "clip": "rect(0 0 0 0)",
                "height": 1,
                "margin": -1,
                "overflow": "hidden",
                "padding": 0,
                "position": "absolute",
                "width": 1,
              }
            }
            type="checkbox"
          />
        </div>
      </ReactSwitch>
      <span
        className="deployment-form-label"
      >
        Enable Metrics
      </span>
    </div>
  </label>
</BooleanParam>
`;
